### PR TITLE
Enable sole dimension attribute

### PR DESCRIPTION
**Pull request by konradha**
This implements the sole `dimension`attribute, e.g.

```fortran
      integer a
      dimension a(1,2)
```

which will then update `a` 's dimension.

### CODE REVIEW

**Comment by certik** [comment]
You have to update tests via `./run_tests.py -u` and commit the result.

**Comment by konradha** [comment]
> You have to update tests via `./run_tests.py -u` and commit the result.

Thanks, yes, I did, but something went wrong.

File: src/lfortran/semantics/ast_common_visitor.h [hunk 1]
```diff
@@ -883,13 +883,56 @@ class CommonVisitor : public AST::BaseVisitor<Derived> {
...
+                for (size_t i=0;i<x.n_syms;++i) { // symbols for line only
+                    AST::var_sym_t &s = x.m_syms[i];
+                    std::string sym = to_lower(s.m_name);
+                    ASR::symbol_t *get_sym = current_scope->get_symbol(sym);
+                    // get actual variable
+                    if (get_sym == nullptr) throw SemanticError("Variable undeclared", x.base.base.loc);
+                    // TODO: check if we can abuse dimension to just update a variables dimensions
+
+                    if (ASR::is_a<ASR::Variable_t>(*get_sym)) {
+                        auto v = (ASR::Variable_t*)get_sym;
```
Comment (certik) [inline src/lfortran/semantics/ast_common_visitor.h:910]:
```suggestion
                        auto v = ASR::down_cast<ASR::Variable_t>(get_sym);
```

File: src/lfortran/semantics/ast_common_visitor.h [hunk 2]
```diff
@@ -883,13 +883,56 @@ class CommonVisitor : public AST::BaseVisitor<Derived> {
...
+                        if (ASR::is_a<ASR::Integer_t>(*(v->m_type))) {
+                            auto actual = (ASR::Integer_t*)(v->m_type);
```
Comment (certik) [inline src/lfortran/semantics/ast_common_visitor.h:912]:
```suggestion
                            ASR::Integer_t* actual = ASR::down_cast<ASR::Integer_t>(v->m_type);
```

File: src/lfortran/semantics/ast_common_visitor.h [hunk 3]
```diff
@@ -883,13 +883,56 @@ class CommonVisitor : public AST::BaseVisitor<Derived> {
...
+                            Vec<ASR::dimension_t> dims;
+                            dims.reserve(al, 0);
+                            process_dims(al, dims, x.m_syms[i].m_dim, x.m_syms[i].n_dim);
+                            v->m_type = (ASR::ttype_t*)LFortran::ASR::make_Integer_t(al, actual->base.base.loc, actual->m_kind, dims.data(), x.m_syms[i].n_dim);
```
Comment (certik) [inline src/lfortran/semantics/ast_common_visitor.h:916]: ASR is constructed to never reuse nodes, so that they can be modified in place:
```suggestion
                            actual->m_dims = dims.data();
                            actual->n_dims = dims.size();
```

**Comment by certik** [comment]
Overall this is the right approach, I left some comments to further simplify it.

**Comment by konradha** [comment]
Thanks. The CI ran with not-yet-updated reference files, this should be fine now. Else I will investigate more.

**Comment by certik** [comment]
I think it's still failing. Make sure you compile, and then run `./run_tests.py`, it should fail just like at the CI. Then you can update the tests.

File: tests/fixedform_global_nestedloop.f [hunk 4]
```diff
@@ -1,8 +0,0 @@
- 20    print *, 1
-       do k=1,10
-        do l=1,10
-            print*,k,l
-        enddo
-      enddo
-      if(0>1) goto 20
-      end
```
Comment (certik) [inline tests/fixedform_global_nestedloop.f:8]: Why was this removed? Don't se need this to work?

**Review by certik — APPROVED** [review]
I think that this is fine. Make sure you either polish the git history or "squash" this before merging.

File: tests/fixedform_global_nestedloop.f (hunk 4 above)
Comment (konradha) [reply tests/fixedform_global_nestedloop.f:8]: Good catch, I think I mistyped.

Comment (certik) [reply tests/fixedform_global_nestedloop.f:8]: Just push a commit to put it back.

**Comment by certik** [comment]
I simplified this in https://github.com/lfortran/lfortran/pull/116.
